File: apps/api/src/foundora/api/sandbox.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from pydantic import BaseModel, ConfigDict, Field

from foundora.api.auth import require_auth, require_csrf
from foundora.auth.service import AuthContext
from foundora.sandbox.service import (
    SandboxConflict,
    SandboxExecutionNotFound,
    SandboxExecutionPage,
    SandboxExecutionRecord,
    SandboxNotCancellable,
    SandboxNotReady,
    SandboxProfileMismatch,
    SandboxQueueUnavailable,
    SandboxService,
)

router = APIRouter(prefix="/sandbox", tags=["generated-code sandbox"])
ExecutionStatus = Literal[
    "requested",
    "waiting_approval",
    "queued",
    "authorizing",
    "running",
    "cleaning",
    "rejected",
    "succeeded",
    "failed",
    "cancelled",
    "timed_out",
    "resource_exhausted",
    "infrastructure_failed",
    "cleanup_failed",
]
GovernanceStatus = Literal[
    "approval_required", "approved", "rejected", "authorized", "denied", "blocked"
]


class RequestSandboxExecution(BaseModel):
    model_config = ConfigDict(extra="forbid")

    idempotency_key: str = Field(
        min_length=8,
        max_length=128,
        pattern=r"^[A-Za-z0-9._:-]+$",
    )


class SandboxApprovalView(BaseModel):
    id: UUID
    status: Literal["pending", "approved", "rejected", "cancelled"]
    prompt: str
    decision_reason: str | None
    requested_at: datetime
    decided_at: datetime | None


class SandboxExecutionSummaryView(BaseModel):
    id: UUID
    business_id: UUID
    website_project_id: UUID
    website_project_version: int
    website_specification_id: UUID
    website_specification_version: int
    profile_id: str
    profile_version: int
    governance_action_id: UUID
    governance_status: GovernanceStatus
    status: ExecutionStatus
    cleanup_status: Literal["pending", "verified", "failed"]
    final_labeled_resource_count: int | None
    cancellation_requested_at: datetime | None
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    updated_at: datetime


class SandboxExecutionDetailView(SandboxExecutionSummaryView):
    harness_contract_version: int
    source_digest: str
    build_digest: str
    source_archive_sha256: str
    source_archive_size_bytes: int
    routes: list[str]
    request_digest: str
    policy_version_id: UUID
    governance_risk_class: Literal["R0", "R1", "R2", "R3", "R4", "R5"]
    governance_rationale: str
    governance_authorized_at: datetime | None
    approval: SandboxApprovalView | None
    worker_recovery_count: int
    attempt_started_at: datetime | None
    heartbeat_at: datetime | None
    runtime_image_id: str | None
    effective_limits: dict[str, object] | None
    effective_limits_digest: str | None
    termination_reason: str | None
    exit_code: int | None
    route_results: list[dict[str, object]] | None
    process_results: dict[str, object] | None
    stdout_excerpt: str | None
    stderr_excerpt: str | None
    stdout_sha256: str | None
    stderr_sha256: str | None
    cleanup_attempts: int
    cleanup_started_at: datetime | None
    cleanup_finished_at: datetime | None
    cleanup_receipt_digest: str | None


class SandboxExecutionPageView(BaseModel):
    business_id: UUID
    executions: list[SandboxExecutionSummaryView]
    total_executions: int
    limit: int
    offset: int


def _summary_view(record: SandboxExecutionRecord) -> SandboxExecutionSummaryView:
    execution = record.execution
    return SandboxExecutionSummaryView(
        id=execution.id,
        business_id=execution.business_id,
        website_project_id=execution.website_project_id,
        website_project_version=execution.website_project_version,
        website_specification_id=execution.website_specification_id,
        website_specification_version=execution.website_specification_version,
        profile_id=execution.profile_id,
        profile_version=execution.profile_version,
        governance_action_id=execution.governance_action_id,
        governance_status=record.action.status,  # type: ignore[arg-type]
        status=execution.status,  # type: ignore[arg-type]
        cleanup_status=execution.cleanup_status,  # type: ignore[arg-type]
        final_labeled_resource_count=execution.final_labeled_resource_count,
        cancellation_requested_at=execution.cancellation_requested_at,
        created_at=execution.created_at,
        started_at=execution.started_at,
        finished_at=execution.finished_at,
        updated_at=execution.updated_at,
    )


def _detail_view(record: SandboxExecutionRecord) -> SandboxExecutionDetailView:
    execution = record.execution
    action = record.action
    approval = record.approval
    summary = _summary_view(record).model_dump()
    return SandboxExecutionDetailView(
        **summary,
        harness_contract_version=execution.harness_contract_version,
        source_digest=execution.source_digest,
        build_digest=execution.build_digest,
        source_archive_sha256=execution.source_archive_sha256,
        source_archive_size_bytes=execution.source_archive_size_bytes,
        routes=execution.routes,
        request_digest=execution.request_digest,
        policy_version_id=execution.policy_version_id,
        governance_risk_class=action.risk_class,  # type: ignore[arg-type]
        governance_rationale=action.rationale,
        governance_authorized_at=action.authorized_at,
        approval=(
            SandboxApprovalView(
                id=approval.id,
                status=approval.status,  # type: ignore[arg-type]
                prompt=approval.prompt,
                decision_reason=approval.decision_reason,
                requested_at=approval.requested_at,
                decided_at=approval.decided_at,
            )
            if approval is not None
            else None
        ),
        worker_recovery_count=execution.worker_recovery_count,
        attempt_started_at=execution.attempt_started_at,
        heartbeat_at=execution.heartbeat_at,
        runtime_image_id=execution.runtime_image_id,
        effective_limits=execution.effective_limits,
        effective_limits_digest=execution.effective_limits_digest,
        termination_reason=execution.termination_reason,
        exit_code=execution.exit_code,
        route_results=execution.route_results,
        process_results=execution.process_results,
        stdout_excerpt=execution.stdout_excerpt,
        stderr_excerpt=execution.stderr_excerpt,
        stdout_sha256=execution.stdout_sha256,
        stderr_sha256=execution.stderr_sha256,
        cleanup_attempts=execution.cleanup_attempts,
        cleanup_started_at=execution.cleanup_started_at,
        cleanup_finished_at=execution.cleanup_finished_at,
        cleanup_receipt_digest=execution.cleanup_receipt_digest,
    )


def _page_view(page: SandboxExecutionPage) -> SandboxExecutionPageView:
    return SandboxExecutionPageView(
        business_id=page.business_id,
        executions=[_summary_view(record) for record in page.executions],
        total_executions=page.total_executions,
        limit=page.limit,
        offset=page.offset,
    )


def _handle(error: Exception) -> HTTPException:
    if isinstance(error, SandboxExecutionNotFound):
        return HTTPException(status_code=404, detail="Sandbox execution not found")
    if isinstance(error, SandboxNotCancellable):
        return HTTPException(status_code=409, detail="Sandbox execution is already terminal")
    if isinstance(error, SandboxConflict):
        return HTTPException(status_code=409, detail=str(error))
    if isinstance(error, (SandboxNotReady, SandboxProfileMismatch)):
        return HTTPException(status_code=422, detail=str(error))
    if isinstance(error, SandboxQueueUnavailable):
        return HTTPException(
            status_code=503,
            detail={"code": "queue_unavailable"},
        )
    return HTTPException(status_code=500, detail="Sandbox operation failed")


@router.get("/executions", response_model=SandboxExecutionPageView)
async def list_sandbox_executions(
    context: Annotated[AuthContext, Depends(require_auth)],
    response: Response,
    limit: Annotated[int, Query(ge=1, le=100)] = 25,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> SandboxExecutionPageView:
    page = await SandboxService().list_executions(context, limit=limit, offset=offset)
    response.headers["Cache-Control"] = "no-store"
    return _page_view(page)


@router.get("/executions/{execution_id}", response_model=SandboxExecutionDetailView)
async def get_sandbox_execution(
    execution_id: UUID,
    context: Annotated[AuthContext, Depends(require_auth)],
    response: Response,
) -> SandboxExecutionDetailView:
    try:
        record = await SandboxService().get_execution(context, execution_id)
    except SandboxExecutionNotFound as error:
        raise _handle(error) from error
    response.headers["Cache-Control"] = "no-store"
    return _detail_view(record)


@router.post(
    "/executions",
    response_model=SandboxExecutionDetailView,
    status_code=status.HTTP_201_CREATED,
)
async def request_sandbox_execution(
    payload: RequestSandboxExecution,
    context: Annotated[AuthContext, Depends(require_csrf)],
    response: Response,
) -> SandboxExecutionDetailView:
    service = SandboxService()
    try:
        execution = await service.request_execution(
            context,
            idempotency_key=payload.idempotency_key,
        )
        record = await service.get_execution(context, execution.id)
    except (SandboxConflict, SandboxNotReady, SandboxProfileMismatch) as error:
        raise _handle(error) from error
    response.headers["Cache-Control"] = "no-store"
    return _detail_view(record)


@router.post(
    "/executions/{execution_id}/start",
    response_model=SandboxExecutionDetailView,
    status_code=status.HTTP_202_ACCEPTED,
)
async def start_sandbox_execution(
    execution_id: UUID,
    context: Annotated[AuthContext, Depends(require_csrf)],
    response: Response,
) -> SandboxExecutionDetailView:
    service = SandboxService()
    try:
        await service.start_execution(context, execution_id)
        record = await service.get_execution(context, execution_id)
    except (
        SandboxConflict,
        SandboxExecutionNotFound,
        SandboxNotReady,
        SandboxProfileMismatch,
        SandboxQueueUnavailable,
    ) as error:
        raise _handle(error) from error
    response.headers["Cache-Control"] = "no-store"
    return _detail_view(record)


@router.post(
    "/executions/{execution_id}/cancel",
    response_model=SandboxExecutionDetailView,
    status_code=status.HTTP_202_ACCEPTED,
)
async def cancel_sandbox_execution(
    execution_id: UUID,
    context: Annotated[AuthContext, Depends(require_csrf)],
    response: Response,
) -> SandboxExecutionDetailView:
    service = SandboxService()
    try:
        await service.cancel_execution(context, execution_id)
        record = await service.get_execution(context, execution_id)
    except (
        SandboxExecutionNotFound,
        SandboxNotCancellable,
        SandboxQueueUnavailable,
    ) as error:
        raise _handle(error) from error
    response.headers["Cache-Control"] = "no-store"
    return _detail_view(record)

File: apps/web/lib/sandbox.ts
import "server-only";

import { cookies } from "next/headers";

export type SandboxExecutionStatus =
  | "requested"
  | "waiting_approval"
  | "queued"
  | "authorizing"
  | "running"
  | "cleaning"
  | "rejected"
  | "succeeded"
  | "failed"
  | "cancelled"
  | "timed_out"
  | "resource_exhausted"
  | "infrastructure_failed"
  | "cleanup_failed";

export interface SandboxExecutionSummary {
  id: string;
  business_id: string;
  website_project_id: string;
  website_project_version: number;
  website_specification_id: string;
  website_specification_version: number;
  profile_id: string;
  profile_version: number;
  governance_action_id: string;
  governance_status: string;
  status: SandboxExecutionStatus;
  cleanup_status: "pending" | "verified" | "failed";
  final_labeled_resource_count: number | null;
  cancellation_requested_at: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  updated_at: string;
}

export interface SandboxExecutionDetail extends SandboxExecutionSummary {
  harness_contract_version: number;
  source_digest: string;
  build_digest: string;
  source_archive_sha256: string;
  source_archive_size_bytes: number;
  routes: string[];
  request_digest: string;
  policy_version_id: string;
  governance_risk_class: string;
  governance_rationale: string;
  governance_authorized_at: string | null;
  approval: {
    id: string;
    status: string;
    prompt: string;
    decision_reason: string | null;
    requested_at: string;
    decided_at: string | null;
  } | null;
  worker_recovery_count: number;
  attempt_started_at: string | null;
  heartbeat_at: string | null;
  runtime_image_id: string | null;
  effective_limits: Record<string, unknown> | null;
  effective_limits_digest: string | null;
  termination_reason: string | null;
  exit_code: number | null;
  route_results: Array<Record<string, unknown>> | null;
  process_results: Record<string, unknown> | null;
  stdout_excerpt: string | null;
  stderr_excerpt: string | null;
  stdout_sha256: string | null;
  stderr_sha256: string | null;
  cleanup_attempts: number;
  cleanup_started_at: string | null;
  cleanup_finished_at: string | null;
  cleanup_receipt_digest: string | null;
}

export interface SandboxExecutionPage {
  business_id: string;
  executions: SandboxExecutionSummary[];
  total_executions: number;
  limit: number;
  offset: number;
}

function isSummary(value: unknown): value is SandboxExecutionSummary {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<SandboxExecutionSummary>;
  return (
    typeof item.id === "string" &&
    typeof item.business_id === "string" &&
    typeof item.website_project_id === "string" &&
    typeof item.website_project_version === "number" &&
    typeof item.governance_action_id === "string" &&
    typeof item.governance_status === "string" &&
    typeof item.status === "string" &&
    typeof item.cleanup_status === "string" &&
    typeof item.created_at === "string" &&
    typeof item.updated_at === "string"
  );
}

function isPage(value: unknown): value is SandboxExecutionPage {
  if (typeof value !== "object" || value === null) return false;
  const page = value as Partial<SandboxExecutionPage>;
  return (
    typeof page.business_id === "string" &&
    Array.isArray(page.executions) &&
    page.executions.every(isSummary) &&
    typeof page.total_executions === "number" &&
    typeof page.limit === "number" &&
    typeof page.offset === "number"
  );
}

function isDetail(value: unknown): value is SandboxExecutionDetail {
  if (!isSummary(value)) return false;
  const detail = value as Partial<SandboxExecutionDetail>;
  return (
    typeof detail.request_digest === "string" &&
    typeof detail.source_digest === "string" &&
    typeof detail.build_digest === "string" &&
    Array.isArray(detail.routes) &&
    typeof detail.governance_rationale === "string" &&
    typeof detail.worker_recovery_count === "number" &&
    typeof detail.cleanup_attempts === "number"
  );
}

async function getSandbox(path: string): Promise<unknown | null> {
  const store = await cookies();
  const session = store.get("id")?.value;
  if (!session) return null;
  try {
    const base = process.env.API_INTERNAL_URL ?? "http://localhost:8000";
    const response = await fetch(`${base}${path}`, {
      cache: "no-store",
      headers: { Cookie: `id=${session}` },
      signal: AbortSignal.timeout(5000),
    });
    if (!response.ok) return null;
    return await response.json();
  } catch {
    return null;
  }
}

export async function getSandboxExecutions(): Promise<SandboxExecutionPage | null> {
  const value = await getSandbox("/sandbox/executions?limit=50&offset=0");
  return isPage(value) ? value : null;
}

export async function getSandboxExecution(
  executionId: string,
): Promise<SandboxExecutionDetail | null> {
  const value = await getSandbox(
    `/sandbox/executions/${encodeURIComponent(executionId)}`,
  );
  return isDetail(value) ? value : null;
}
